transformer_engine/pytorch/attention.py

```diff
@@ -439,9 +440,17 @@ def forward(ctx, q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k
         recv_src = cp_global_ranks[(rank + cp_size - 1) % cp_size]
         batch_p2p_comm = int(os.getenv("NVTE_BATCH_MHA_P2P_COMM", "0")) or (cp_size == 2)
 
-        # [b, s, np, hn] -> [b, 2, s//2, np, hn]
-        q, k, v = [x.view(x.shape[0], 2, x.shape[1]//2, *x.shape[2:]) for x in [q, k, v]]
+        causal = (attn_mask_type == "causal")
+
+        if causal:
+            # [b, s, np, hn] -> [b, 2, s//2, np, hn]
+            q, k, v = [x.view(x.shape[0], 2, x.shape[1]//2, *x.shape[2:]) for x in [q, k, v]]
```

Comment:
Hi @xrennvidia are you trying to load balance causal attention att(Q, K_i, V_i). If that is true, I have done this for both fwd and bwd with triton kernel, and there is no need for users to take care  of it outside of kernels:

https://github.com/yiakwy-xpu-ml-framework-team/triton/blob/add_support_flash_attention_v3/python/triton/ops/flash_attention.py